### PR TITLE
[spirv] Move external arrays into seperate buffers

### PR DESCRIPTION
**Pull request by bobcao3**
Related issue = https://github.com/taichi-dev/taichi/issues/3642

First step towards ndarray on SPIRV backends.

Also fixed a few small things to speed up compilation a small bit.

cc: @xwang186

### CODE REVIEW

File: taichi/backends/vulkan/runtime.cpp [hunk 1]
```diff
@@ -427,9 +441,24 @@ void VkRuntime::launch_kernel(KernelHandle handle, RuntimeContext *host_ctx) {
       &ti_kernel->ti_kernel_attribs().ctx_attribs, host_ctx, device_,
       host_result_buffer_, ctx_buffer.get(), ctx_buffer_host.get());
 
+  std::unordered_map<int, DeviceAllocation> ext_arrays;
```
Comment (ailzhang) [inline taichi/backends/vulkan/runtime.cpp:444]: FYI in opengl we don't redo device allocation if the size doesn't change. maybe it helps with perf a little bit? https://github.com/taichi-dev/taichi/blob/master/taichi/backends/opengl/opengl_api.cpp#L400 (can be a followup ofc)

**Review by ailzhang — APPROVED** [review]
Awesome thanks!

File: taichi/program/program.h [hunk 2]
```diff
@@ -5,6 +5,7 @@
 #include <functional>
 #include <optional>
 #include <atomic>
+#include <stack>
```
[the remark this answers is not in the document]
Comment (bobcao3) [reply taichi/program/program.h:8]: It's used in this class

File: taichi/backends/vulkan/runtime.cpp (hunk 1 above)
Comment (bobcao3) [reply taichi/backends/vulkan/runtime.cpp:444]: Let's leave it for later, the balance between memory usage and allocation performance is hard

File: taichi/backends/vulkan/runtime.cpp [hunk 3]
```diff
@@ -167,8 +178,8 @@ class HostDeviceContextBlitter {
       const auto dt = ret.dt;
       do {
         if (ret.is_array) {
-          void *host_ptr = host_ctx_->get_arg<void *>(i);
-          std::memcpy(host_ptr, device_ptr, ret.stride);
+          // void *host_ptr = host_ctx_->get_arg<void *>(i);
+          // std::memcpy(host_ptr, device_ptr, ret.stride);
```
Comment (xwang186) [inline taichi/backends/vulkan/runtime.cpp:182]: We could remove useless comments in the codes

File: taichi/codegen/spirv/kernel_utils.h [hunk 4]
```diff
@@ -20,16 +20,11 @@ namespace spirv {
  * Per offloaded task attributes.
  */
 struct TaskAttributes {
-  enum class BufferType {
-    Root,
-    GlobalTmps,
-    Context,
-    ListGen,
-  };
+  enum class BufferType { Root, GlobalTmps, Context, ListGen, ExtArr };
```
Comment (k-ye) [inline taichi/codegen/spirv/kernel_utils.h:23]: OOC, do we really support ListGen buffer in SPIR-V codegen..?

**Review by k-ye — APPROVED** [review]
LGTM! Please resolve @xwang186 's comment :-)

File: taichi/codegen/spirv/kernel_utils.h (hunk 4 above)
Comment (bobcao3) [reply taichi/codegen/spirv/kernel_utils.h:23]: We have very low level all dense->bitmasked support

**Pull request merged**